frontend/src/metabase-lib/expressions/diagnostics.ts

```diff
@@ -151,8 +167,26 @@ function prattCompiler(source, startRule, legacyQuery, name) {
     }
   } catch (err) {
     console.warn("compile error", err);
-    return err;
+
+    if (isErrorWithMessage(err)) {
+      return err;
+    }
+
+    return { message: t`Invalid expression` };
   }
 
   return null;
 }
+
+function isBooleanExpression(
+  expr: unknown,
+): expr is [string, ...Expr[]] & { node?: Node } {
+  return (
+    Array.isArray(expr) &&
+    (LOGICAL_OPS.includes(expr[0]) || COMPARISON_OPS.includes(expr[0]))
+  );
+}
+
+function isErrorWithMessage(err: unknown): err is ErrorWithMessage {
+  return err instanceof Error && typeof err.message === "string";
```

Comment:
We should either make the guard less strict or make `ErrorWithMessage` extend `Error`.
But probably less strict, because `prattCompiler` now returns something that is not an instance of Error, i.e.
```
{ message: t`Invalid expression` }
``` 

```suggestion
  return typeof err === 'object' && !err && typeof err.message === "string";
```